crates/signature-mpc/src/twopc_mpc_protocols/mod.rs

```
@@ -408,23 +488,103 @@ pub fn decrypt_signature_decentralized_party_sign(
+ ///
+ /// In case one of the involved parties failed to generate a proof, only this party will be returned.
+ /// If all the involved parties sent a proof, and some of the proofs are invalid, all the parties that
+ /// their proof was invalid will be returned.
+pub fn identify_message_malicious_parties(
+    verification_round_party: SignaturePartialDecryptionProofVerificationParty,
+    partial_signature_decryption_shares: HashMap<PartyID, DecryptionShare>,
+    masked_nonce_decryption_shares: HashMap<PartyID, DecryptionShare>,
+    decryption_key_share_public_parameters: DecryptionPublicParameters,
+    signature_partial_decryption_proofs: HashMap<PartyID, PartialDecryptionProof>,
+    decrypters: Vec<PartyID>,
+) -> Vec<PartyID>
+{
+    let lagrange_coefficients: HashMap<PartyID, AdjustedLagrangeCoefficientSizedNumber> = decrypters
+        .clone()
+        .into_iter()
+        .map(|j| {
+            (
+                j,
+                DecryptionKeyShare::compute_lagrange_coefficient(
+                    j,
+                    decryption_key_share_public_parameters.number_of_parties,
+                    decrypters.clone(),
+                    &decryption_key_share_public_parameters,
+                ),
+            )
+        })
+        .collect();
 
-        let signature_s_inner: k256::Scalar = signature_s.into();
+    let error = verification_round_party.identify_malicious_decrypters(
+        lagrange_coefficients,
+        partial_signature_decryption_shares,
+        masked_nonce_decryption_shares,
+        signature_partial_decryption_proofs,
+        &mut OsRng,
+    );
+
+    match error {
+        Error::Tiresias(tiresias::Error::ProtocolError(
+                            ProtocolError::ProofVerificationError {
+                                malicious_parties,
+                                ..
+                            })) => malicious_parties,
+        _ => Vec::new(),
```

What is this?

---

Probably we want to report an internal error here